### PR TITLE
rzo gem (bundle exec rizzo --help)

### PR DESCRIPTION
**Pull request by jeffmccune**
Refactor existing rizzo code into a Ruby gem, suitable for including in a
control repo Gemfile.

### CODE REVIEW

File: lib/rizzo/version.rb
Comment (ghoneycutt) [inline lib/rizzo/version.rb]: why is this here and in `lib/rizzo.rb` ?

**Comment by ghoneycutt** [comment]
Have not heard back from the maintainer of the rizzo rubygem, so grabbed the name `rzo`.

https://rubygems.org/gems/rzo

**Comment by ghoneycutt** [comment]
for our first version, we just need the functionality of writing out the Vagrantfile from a template to the first (top) layer control repo's directory.

**Comment by jeffmccune** [comment]
Got it, should have that implemented by Monday.

File: lib/rizzo/version.rb
Comment (jeffmccune) [reply lib/rizzo/version.rb]: It should just be in the version.

File: README.md
Comment (jeffmccune) [inline README.md]: This is intended to be the spec_helper of the top level control repo, yes?

Comment (ghoneycutt) [reply README.md]: yeah, as part of the effort to allow for ensuring that we can compile catalogs for each role

**Comment by jeffmccune** [comment]
@ghoneycutt Would love to get your feedback on this, the `Vagrantfile` generation has been implemented.

What do you think about adding `rzo` as a Gem dependency inside Puppetdata's Gemfile?  Here's how it works now:

First, add the dependency to `puppetdata/Gemfile`

```diff
diff --git a/Gemfile b/Gemfile
index 572f681..5c2f3b0 100644
--- a/Gemfile
+++ b/Gemfile
@@ -20,6 +20,7 @@ gem 'puppet-lint-undef_in_function-check'
 gem 'puppet-lint-unquoted_string-check'
 gem 'puppet-lint-variable_contains_upcase'
 gem 'puppet-syntax'
+gem 'rzo', git: 'https://github.com/jeffmccune/rizzo.git', branch: 'turn_into_gem'

 gem 'serverspec', :require => false
 gem 'vagrant-wrapper', :require => false
```

Install the gems with `bundle install --path .bundle/gem/`

Then generate the Vagrantfile:

```
bundle exec rzo generate > Vagrantfile
```

Then boot the VM's as normal:

```
vagrant up infra-puppetfourca
```

Here's what is currently implemented:

```
be rizzo --help
usage: rizzo [GLOBAL OPTIONS] SUBCOMMAND [ARGS]
Sub Commands:

  generate     Initialize Vagrantfile in top control repo

Global options: (Note, command line arguments supersede ENV vars in {}'s)
  -l, --logto=<s>     Log file to write to or keywords STDOUT, STDERR {RZO_LOGTO} (default: STDERR)
  -s, --syslog        Log to syslog
  -v, --verbose       Set log level to INFO
  -d, --debug         Set log level to DEBUG
  -c, --config=<s>    Rizzo config file {RZO_CONFIG} (default: ~/.rizzo.json)
  -e, --version       Print version and exit
  -h, --help          Show this message
```

```
bundle exec rizzo generate --help
rizzo generate options:
  -o, --vagrantfile=<s>    Output Vagrantfile (default: STDOUT)
  -h, --help               Show this message
```

**Comment by ghoneycutt** [comment]
`rzo generate` should create the Vagrantfile instead of having to run `bundle exec rzo generate > Vagrantfile`

File: .travis.yml
Comment (ghoneycutt) [inline .travis.yml:9]: Which ruby versions should we support? 2.1.9 is what puppet4 uses and puppet5 is currently on 2.4.1. Should we support both? others?

Comment (jeffmccune) [reply .travis.yml:9]: Let's support the same ruby version shipping in the AIO package of the Puppet Agent the customer support.  So we can start with Puppet 4, Puppet 5 with MRI 2.1.9 and 2.4.1 and add future MRI versions for Puppet 6 onward.

Comment (ghoneycutt) [reply .travis.yml:9]: Cool. Let's change this to test against 2.1.9 and 2.4.1.

File: .travis.yml
Comment (ghoneycutt) [inline .travis.yml]: Should not have to install bundler as it will be there when you specify the lang as ruby.

File: Gemfile
Comment (ghoneycutt) [inline Gemfile]: what does this do?

File: README.md
Comment (ghoneycutt) [inline README.md]: This functionality can wait until the next release. The first release should just handle generating the Vagrantfile

File: spec/rzo_spec.rb
Comment (ghoneycutt) [inline spec/rzo_spec.rb]: we could have a couple `.rizzo.json` files and see that the generated Vagrantfile matches a fixture.

File: rzo.gemspec
Comment (ghoneycutt) [inline rzo.gemspec]: I think this should be `Apache-2.0` to conform with SPDX

File: rzo.gemspec
Comment (ghoneycutt) [inline rzo.gemspec]:
```ruby
spec.authors = ['Garrett Honeycutt', 'Jeff McCune']
```

File: rzo.gemspec
Comment (ghoneycutt) [inline rzo.gemspec]: `code@learnpuppet.com`

File: Gemfile
Comment (jeffmccune) [reply Gemfile]: No idea, it was part of the skeleton generated by `bundle gem rzo`.  I'll remove it.

File: .travis.yml
Comment (jeffmccune) [reply .travis.yml]: Good to know, it was part of the skeleton generated by `bundle gem rzo`.  I'll remove it.

File: .travis.yml
Comment (jeffmccune) [reply .travis.yml:9]: Will do.

File: rzo.gemspec
Comment (ghoneycutt) [inline rzo.gemspec]: `Rizzo (rzo) is a tool for working with Vagrant and layered Puppet control repos`

**Comment by jeffmccune** [comment]
> rzo generate should create the Vagrantfile instead of having to run bundle exec rzo generate > Vagrantfile

Fixed, generate writes to Vagrantfile by default, allows writing to STDOUT

File: rzo.gemspec
Comment (jeffmccune) [reply rzo.gemspec]: Fixed

File: rzo.gemspec
Comment (jeffmccune) [reply rzo.gemspec]: Fixed.  FYI, also added myself as an author.

File: rzo.gemspec
Comment (jeffmccune) [reply rzo.gemspec]: Fixed

File: spec/rzo_spec.rb
Comment (jeffmccune) [reply spec/rzo_spec.rb]: Will add and make travis happy.

File: README.md
Comment (jeffmccune) [reply README.md]: Roger, wilco.

File: README.md
Comment (jeffmccune) [reply README.md]: Going to push the spec_helper changes to a subsequent release unless you want it in this PR.

**Comment by jeffmccune** [comment]
This is good to go.  Only thing missing mentioned in the review is some end-to-end testing given .rizzo.json inputs asserting against the Vagrantfile output.

File: rzo.gemspec
Comment (jeffmccune) [reply rzo.gemspec]: Fixed

**Comment by jeffmccune** [comment]
End-to-end tests have been added.  This should be ready for merge.

File: Gemfile
Comment (ghoneycutt) [reply Gemfile]: The above still needs to be removed

File: lib/rzo/trollop.rb
Comment (ghoneycutt) [inline lib/rzo/trollop.rb:5]: Curious why this code is copy/pasted here instead of us calling out to a library in the trollop gem

File: spec/fixtures/_complete_rizzo.json
Comment (ghoneycutt) [inline spec/fixtures/_complete_rizzo.json]: all of these names seem a little too close to home. Let's just pick a few nodes with names like puppet, web, app and db with hostnames like puppet.dev.example.com

File: rzo.gemspec
Comment (ghoneycutt) [reply rzo.gemspec]: i dont see the email / author updates

**Comment by ghoneycutt** [comment]
The README should also be cleaned up so it just documents what the program does and removes my ideas on what it could do.

Would be nice to put the output of `rzo --help` in the README, too.

File: Gemfile
Comment (jeffmccune) [reply Gemfile]: Removed.

File: rzo.gemspec
Comment (jeffmccune) [reply rzo.gemspec]: Hrm, a patch seems to have gotten lost in my rebase.  Fixing this again now.

File: rzo.gemspec
Comment (jeffmccune) [reply rzo.gemspec]: Fixed for real.

File: rzo.gemspec
Comment (jeffmccune) [reply rzo.gemspec]: Fixed for real.

File: lib/rzo/trollop.rb
Comment (jeffmccune) [reply lib/rzo/trollop.rb:5]: It's based on my experience working with Trollop in Puppet core.  Trollop is designed to be vendored, and it avoids issues in the situation two different command line utilities depend on different versions of Trollop.

File: spec/fixtures/_complete_rizzo.json
Comment (jeffmccune) [reply spec/fixtures/_complete_rizzo.json]: Fixed.

File: rzo.gemspec
Comment (ghoneycutt) [inline rzo.gemspec]: doesn't this have to be removed so I can release the gem?

Comment (jeffmccune) [reply rzo.gemspec]: I'm not 100% sure, but it definitely looks that way.  I set the allowed push host to `https://rubygems.org`.

File: README.md
Comment (ghoneycutt) [inline README.md]: readme needs cleaned up

File: lib/rzo/app/templates/Vagrantfile.erb
Comment (ghoneycutt) [inline lib/rzo/app/templates/Vagrantfile.erb]: A line with the version of rzo, something like `This file was generated with Rizzo v0.1.0 - https://github.com/ghoneycutt/rizzo`

File: lib/rzo/app/templates/environment.conf.erb
Comment (ghoneycutt) [inline lib/rzo/app/templates/environment.conf.erb]: same generated by line as in the Vagrantfile

File: lib/rzo/trollop.rb
Comment (ghoneycutt) [reply lib/rzo/trollop.rb:5]: Makes sense. Could you post a sentence at the top of the file that explains why this was vendored.

**Review by ghoneycutt — CHANGES_REQUESTED** [review]

File: lib/rzo/app/templates/environment.conf.erb
Comment (jeffmccune) [reply lib/rzo/app/templates/environment.conf.erb]: I'm actually going to remove this template, the environment.conf is inlined in the generated Vagrantfile.  For example:

```ruby
config.vm.provision 'shell', inline: "echo 'modulepath = ./modules:./puppetdata/modules:./ghoneycutt/modules' > /tmp/bootstrap_puppet4/environment.conf"
```

This was a lot easier than managing different environment files for different VM types.

File: lib/rzo/app/templates/Vagrantfile.erb
Comment (jeffmccune) [reply lib/rzo/app/templates/Vagrantfile.erb]: Added.  Looks like:

```
# This file generated with Rizzo 0.1.0 on 2017-08-18 13:00:09 -0700
# https://github.com/ghoneycutt/rizzo
Vagrant.configure(2) do |config|
```

File: README.md
Comment (jeffmccune) [reply README.md]: Readme has been cleaned up

File: lib/rzo/app/templates/environment.conf.erb
Comment (ghoneycutt) [reply lib/rzo/app/templates/environment.conf.erb]: perhaps it should just do this within the puppetmaster, so `/etc/puppetlabs/code/environment.conf` or wherever it is on the host.

Comment (jeffmccune) [reply lib/rzo/app/templates/environment.conf.erb]: The rizzo generate command adds that inline provision shell command only for puppetmaster nodes.  All the other nodes have the line omitted.

Is that what you meant?

Comment (jeffmccune) [reply lib/rzo/app/templates/environment.conf.erb]: Just as an FYI too, the bootstrap script installs the file using:

```shell
install -o root -g puppet -m 0644 ${BASE_DIR}/environment.conf ${PUPPET_ENVS}/${ENVIRONMENT}/environment.conf
```

This evaluates to:

```
install -o root -g puppet -m 0644 \
  /tmp/bootstrap_puppet4/environment.conf \
  /etc/puppetlabs/code/environments/production/environment.conf
```

**Review by ghoneycutt — APPROVED** [review]

**Pull request merged**
